### PR TITLE
Pre cache more peers than max limits

### LINKED ISSUE #972: Pre-cache more peers than max limit during during discovery
# Problem

Current discovery process on a mobile device will stop itself when max amount of peers is found.
It is still possible that we will get +1-2 than max peers in our cache because we rebalance peers after confirmation that the got connected.
It would be a good optimization to cache more than max amount of peers before stopping discovery, so 
that when we loose connection with an initial set of peers we can set up new connections immediately even without turning on discovery.

# Implementation

TBD

# Acceptance Criteria

Cache more peers than max limit, x2 or +4, can be decided during implementation.
It is also important that we stop discovery eventually.

### PR DESCRIPTION
**Pull request by adriacidre**
This feature allows discovery process to cache more peers (double) than those defined on `maxPeers`.
Basically, the topic search won't be stopped when maxPeers is reached, but will be when `maxPeers * 2` is, or, on a timeout of 10 seconds since `maxPeers` was reached.

Important changes:
- [x] Introduced maxCachedPeers so the discovery search keeps running after max connections is reached and until it reaches max cached connections.
- [x] If DefaultTopicStopSearchDelay is set up, StopSearch will be delayed.

Closes #972

### CODE REVIEW

**Comment by adriacidre** [comment]
@dshulyak Do you mind having a look at this changes? I'm not sure if having peers cached on `pendingPeers` makes sense or they should be stored somewhere else.

**Comment by dshulyak** [comment]
the goal is to get more peers into the persistent cache, because mobile app is restarted quite frequently.
if i remember it correctly, we add a peer to persistent cache when we are able to establish a connection with such peer, but we already had max peers for a topic at the time. i think that we can use this behavior to validate every connection and then count how much peers we actually have in persistent cache.

another question is how long we want to wait for such backup connections. discovery can generate a significant amount of traffic, which is undesirable on mobile. we already have a hard stop implemented for peer pool, maybe we should adjust it once we found max limit of peers, to 10s for example.

**Comment by adriacidre** [comment]
Flaky unit test: 
```
INFO [05-28|12:02:04] RLPx listener up                         self="enode://535d827b30ee23d76389a38faf12697519d5fbc70b5a8896bc47f5ea31b553c8b05c5916d9facef2860cf0d04f776ad5b00062bc61f09b675a3f7f43edf7119d@[::]:33739?discport=0"
--- FAIL: TestPeerPoolSimulationSuite (20.17s)
    --- PASS: TestPeerPoolSimulationSuite/TestPeerPoolCache (0.08s)
    --- FAIL: TestPeerPoolSimulationSuite/TestSingleTopicDiscoveryWithFailover (20.08s)
	Error Trace:	peerpool_test.go:101
			peerpool_test.go:190
	Error:      	timed out waiting for a peer
	Error Trace:	peerpool_test.go:191
	Error:      	Not equal: 
	            	expected: discover.HexID("88cd4a7b4805a8b5d123ef7ccb3916ab538efb97026c586a775cc60c1dd7e2a4c86b22407ac5b4677b59eef24d8850d1c32691c3df1e3537f25c0f95d3641ce7")
	            	actual: discover.HexID("00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000")
	            	
	            	Diff:
	            	--- Expected
	            	+++ Actual
	            	@@ -1,2 +1,2 @@
	            	-(discover.NodeID) (len=64) 88cd4a7b4805a8b5d123ef7ccb3916ab538efb97026c586a775cc60c1dd7e2a4c86b22407ac5b4677b59eef24d8850d1c32691c3df1e3537f25c0f95d3641ce7
	            	+(discover.NodeID) (len=64) 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
	            	 
	Error Trace:	peerpool_test.go:112
			peerpool_test.go:193
	Error:      	timed out waiting a pool event
=== RUN   TestPeerPoolMaxPeersOverflow
INFO [05-28|12:02:24] Starting P2P networking 
```

File: geth/peers/peerpool.go [hunk 1]
```diff
@@ -222,37 +229,64 @@ func (p *PeerPool) handleServerPeers(server *p2p.Server, events <-chan *p2p.Peer
 				retryDiscv5 = time.After(discoveryRestartTimeout)
 				log.Error("starting discv5 failed", "error", err, "retry", discoveryRestartTimeout)
 			}
+		case <-stopDiscv5:
+			p.handleStopTopics(server)
+			SendDiscoverySummary(server.PeersInfo())
```
Comment (adambabik) [inline geth/peers/peerpool.go]: `SendDiscoverySummary` should be sent immediately when a number of peers changes. Why not leave it in `case p2p.PeerEventTypeAdd:`?

File: geth/peers/topicpool.go [hunk 2]
```diff
@@ -141,6 +149,40 @@ func (t *TopicPool) BelowMin() bool {
 	return len(t.connectedPeers) < t.limits.Min
 }
 
+// MaxCachedReached returns true if max number of cached peers is reached.
+func (t *TopicPool) MaxCachedReached() bool {
```
Comment (adambabik) [inline geth/peers/topicpool.go]: Why is it exported?

File: geth/peers/topicpool.go [hunk 3]
```diff
@@ -141,6 +149,40 @@ func (t *TopicPool) BelowMin() bool {
...
+	if t.maxCachedPeers == 0 {
+		return true
+	}
+	peers := t.cache.GetPeersRange(t.topic, 100)
```
Comment (adambabik) [inline geth/peers/topicpool.go]: What of `t.maxCachedPeers` is greater than `100`? It's better to use `GetPeersRange(t.topic, t.maxCachedPeers)`.

File: geth/peers/peerpool.go [hunk 4]
```diff
@@ -222,37 +229,64 @@ func (p *PeerPool) handleServerPeers(server *p2p.Server, events <-chan *p2p.Peer
...
+// peers are stopped.
+func (p *PeerPool) handleStopTopics(server *p2p.Server) {
+	if p.opts.AllowStop {
+		for _, t := range p.topics {
+			if t.readyToStopSearch() {
+				t.StopSearch()
+			}
+		}
+	}
+	if p.allTopicsReadyToBeStopped() {
```
Comment (adambabik) [inline geth/peers/peerpool.go]: Shouldn't we check if actually each topic is stopped? It seems strange that we just check if topics are ready to be stopped but we do not stop them. It's done above of course, so all good, just it's strange when you read the code.

File: geth/peers/peerpool.go [hunk 5]
```diff
@@ -222,37 +229,64 @@ func (p *PeerPool) handleServerPeers(server *p2p.Server, events <-chan *p2p.Peer
...
+			t.setStopSearchTimeout(p.opts.TopicStopSearchDelay)
+		}
+	}
+}
+
+// handleStopTopics stops the search on any topics having reached its max cached
+// limit or its delay stop is expired, additionally will stop discovery if all
+// peers are stopped.
+func (p *PeerPool) handleStopTopics(server *p2p.Server) {
+	if p.opts.AllowStop {
```
Comment (adambabik) [inline geth/peers/peerpool.go]: Wouldn't it be possible to do:
```go
if !p.opts.AllowStop {
  return
}
```

File: geth/peers/topicpool_test.go [hunk 6]
```diff
@@ -268,3 +268,37 @@ func (s *TopicPoolSuite) TestPeerDontAddTwice() {
...
+	s.topicPool.maxCachedPeers = 3
+	peer1 := discv5.NewNode(discv5.NodeID{1}, s.peer.Self().IP, 32311, 32311)
+	peer2 := discv5.NewNode(discv5.NodeID{2}, s.peer.Self().IP, 32311, 32311)
+	peer3 := discv5.NewNode(discv5.NodeID{3}, s.peer.Self().IP, 32311, 32311)
+	s.topicPool.processFoundNode(s.peer, peer1)
+	s.topicPool.processFoundNode(s.peer, peer2)
+	s.topicPool.processFoundNode(s.peer, peer3)
+	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer1.ID))
+	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer2.ID))
+	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer3.ID))
```
Comment (adambabik) [inline geth/peers/topicpool_test.go:290]: The logic will call `removeServerPeer` immediately when the limit is reached. I would call `ConfirmDropped()` for `peer1` and `peer2` here.

**Comment by adambabik** [comment]
Overall, LGTM but I am worried about the complexity of this package. We can introduce this change but we need to refactor the package using a state machine and some design patterns like adapter ASAP. Each feature like this should be encapsulated from the main logic.

File: geth/peers/peerpool.go (hunk 1 above)
Comment (adriacidre) [reply geth/peers/peerpool.go]: Since stopping search happens asynchronously (and also the change on number of peers) i moved `SendDiscoverySummary` where this change on the number of peers is happening. 

Additionally some tests were broken as `discovery.stop` was happening just after `discovery.summary`, so I kept the same expected behaviour here.

File: geth/peers/peerpool.go (hunk 4 above)
Comment (adriacidre) [reply geth/peers/peerpool.go]: yep, its also done some lines above, i'll check stopped instead.

File: geth/peers/peerpool.go (hunk 1 above)
Comment (adambabik) [reply geth/peers/peerpool.go]: > moved SendDiscoverySummary where this change on the number of peers is happening.

But it does not happen here. It happens exactly when we receive an event from `events` channel. This block is called after `p.opts.TopicStopSearchDelay` which may be arbitrary. Sending this signal after a few seconds since adding a peer does not make sense and it's pretty confusing.

**Comment by adriacidre** [comment]
@adambabik agree, current logic it's really hard to follow, already seen some comments about it:
```
// @TODO(adam): split it into peers and discovery management loops. This should
// simplify the whole logic and allow to remove `timeout` field from `PeerPool`.
```
If this is a priority, I can start looking at how to improve this package just after merging the current feature.

BTW all other suggestions are implemented now :)

**Comment by adambabik** [comment]
@adriacidre it's not confirmed that this suggestion is gonna work. Overall, our goal should be logic that clearly communicates in what state the peer and topic pools are and has explicit transitions between the states, thus, a simple state machine seems like the best option. Take a look at this video: https://www.youtube.com/watch?v=HxaD_trXwRE and slides https://talks.golang.org/2011/lex.slide#1. That would be my approach first. 

<blockquote><div> YouTube</div><div><strong><a href="https://www.youtube.com/watch?v=HxaD_trXwRE">Lexical Scanning in Go - Rob Pike</a></strong></div><div>Lexical Scanning in Go, a talk by Rob Pike at Google Technology User Group given on Tuesday, 30 August 2011. The slides: http://rspace.googlecode.com/hg/slid...</div></blockquote>
<blockquote><div><strong><a href="https://talks.golang.org/2011/lex.slide#1">Lexical Scanning in Go</a></strong></div></blockquote>

**Comment by dshulyak** [comment]
i agree that it grew quite fast into an overly complex thing, but it also seems that some features can be removed.

for example:
1. using heap for topics is not the best strategy, the better would be to get a bunch of locally cached peers and then rebalance in case if we got more connections than needed. and it actually will make codebase simpler
2. from my pov we don't need a timeout for moving from fast to slow sync, only timeout for hardstop
3. maybe we don't even need  such thing as max limit, just start fast until we get min limit and then turn slow until hardstop. the goal would be to keep traffic under control and get as much valuable peers as possible.

so i would to rethink use cases before refactoring this codebase into proposed state machine, it might be helpful but not that necessary

**Comment by adriacidre** [comment]
Flaky tests report:

1.- Random data race on unit tests
```
=== RUN   TestZeroSubsciptionsOptimization
==================
WARNING: DATA RACE
Read at 0x00c4200261e0 by goroutine 14:
  github.com/status-im/status-go/services/rpcfilters.TestZeroSubsciptionsOptimization()
      /home/travis/gopath/src/github.com/status-im/status-go/services/rpcfilters/latest_block_changed_event_test.go:75 +0x672
  testing.tRunner()
      /home/travis/.gimme/versions/go1.10.1.linux.amd64/src/testing/testing.go:777 +0x16d
Previous write at 0x00c4200261e0 by goroutine 15:
  github.com/status-im/status-go/services/rpcfilters.TestZeroSubsciptionsOptimization.func1()
      /home/travis/gopath/src/github.com/status-im/status-go/services/rpcfilters/latest_block_changed_event_test.go:44 +0x98
  github.com/status-im/status-go/services/rpcfilters.latestBlockProviderTest.GetLatestBlock()
      /home/travis/gopath/src/github.com/status-im/status-go/services/rpcfilters/latest_block_changed_event_test.go:19 +0x8e
  github.com/status-im/status-go/services/rpcfilters.(*latestBlockChangedEvent).Start.func1()
      /home/travis/gopath/src/github.com/status-im/status-go/services/rpcfilters/latest_block_changed_event.go:88 +0x1d9
Goroutine 14 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.10.1.linux.amd64/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      /home/travis/.gimme/versions/go1.10.1.linux.amd64/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      /home/travis/.gimme/versions/go1.10.1.linux.amd64/src/testing/testing.go:777 +0x16d
  testing.runTests()
      /home/travis/.gimme/versions/go1.10.1.linux.amd64/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.10.1.linux.amd64/src/testing/testing.go:978 +0x2cd
  main.main()
      _testmain.go:50 +0x22a
Goroutine 15 (running) created at:
  github.com/status-im/status-go/services/rpcfilters.(*latestBlockChangedEvent).Start()
      /home/travis/gopath/src/github.com/status-im/status-go/services/rpcfilters/latest_block_changed_event.go:80 +0x130
  github.com/status-im/status-go/services/rpcfilters.TestZeroSubsciptionsOptimization()
      /home/travis/gopath/src/github.com/status-im/status-go/services/rpcfilters/latest_block_changed_event_test.go:52 +0x3e7
  testing.tRunner()
      /home/travis/.gimme/versions/go1.10.1.linux.amd64/src/testing/testing.go:777 +0x16d
==================
--- FAIL: TestZeroSubsciptionsOptimization (0.03s)
	testing.go:730: race detected during execution of test
```

2.- Failing transaction
```
INFO [05-29|08:59:30] Notification received                    package=status-go/signal                    event="{\"type\":\"node.stopped\",\"event\":null}"
--- FAIL: TestJailRPCTestSuite (122.53s)
    --- FAIL: TestJailRPCTestSuite/TestJailVMPersistence (15.96s)
    	jail_rpc_test.go:362: CALL START: ["sendTestTx"] {"amount": "0.000001", "from": "0xF35E0325dad87e2661c4eF951d58727e6d583d5c"}
    	jail_rpc_test.go:367: CALL END: ["sendTestTx"] {"amount": "0.000001", "from": "0xF35E0325dad87e2661c4eF951d58727e6d583d5c"}
    	jail_rpc_test.go:343: Transaction queued (will be completed shortly): {id: 530461ff-8018-4b79-a101-e1acfc396fd9}
    	jail_rpc_test.go:362: CALL START: ["sendTestTx"] {"amount": "0.000002", "from": "0xF35E0325dad87e2661c4eF951d58727e6d583d5c"}
    	jail_rpc_test.go:367: CALL END: ["sendTestTx"] {"amount": "0.000002", "from": "0xF35E0325dad87e2661c4eF951d58727e6d583d5c"}
    	jail_rpc_test.go:343: Transaction queued (will be completed shortly): {id: 03e9aad1-0e19-444e-b069-848b842ab046}
    	jail_rpc_test.go:362: CALL START: ["ping"] {"pong": "Ping1", "amount": 0.42}
    	jail_rpc_test.go:367: CALL END: ["ping"] {"pong": "Ping1", "amount": 0.42}
    	jail_rpc_test.go:362: CALL START: ["ping"] {"pong": "Ping2", "amount": 0.42}
    	jail_rpc_test.go:367: CALL END: ["ping"] {"pong": "Ping2", "amount": 0.42}
    	jail_rpc_test.go:397: 0.84
	Error Trace:	jail_rpc_test.go:398
	Error:      	Max difference between 0.840003 and 0.84 allowed is 1e-07, but difference was 3.000000000086267e-06
	Error Trace:	jail_rpc_test.go:348
			signals.go:77
			asm_amd64.s:573
			cgocall.go:316
			cgocall.go:194
			asm_amd64.s:826
			asm_amd64.s:673
			cgocall.go:131
			signals.go:41
			events_sign.go:20
			notifications.go:32
			asm_amd64.s:2361
	Error:      	Received unexpected error:
	            	gas required exceeds allowance or always failing transaction
	Messages:   	cannot complete queued transaction[03e9aad1-0e19-444e-b069-848b842ab046]: gas required exceeds allowance or always failing transaction
    	jail_rpc_test.go:352: Transaction complete: https://rinkeby.etherscan.io/tx/0x0000000000000000000000000000000000000000000000000000000000000000
	Error Trace:	jail_rpc_test.go:348
			signals.go:77
			asm_amd64.s:573
			cgocall.go:316
			cgocall.go:194
			asm_amd64.s:826
			asm_amd64.s:673
			cgocall.go:131
			signals.go:41
			events_sign.go:20
			notifications.go:32
			asm_amd64.s:2361
	Error:      	Received unexpected error:
	            	Client is shutting down
	Messages:   	cannot complete queued transaction[530461ff-8018-4b79-a101-e1acfc396fd9]: Client is shutting down
    	jail_rpc_test.go:352: Transaction complete: https://rinkeby.etherscan.io/tx/0x0000000000000000000000000000000000000000000000000000000000000000
```

cc: @adambabik

File: geth/peers/topicpool.go [hunk 7]
```diff
@@ -141,6 +149,40 @@ func (t *TopicPool) BelowMin() bool {
...
+// been stopped before.
+func (t *TopicPool) setStopSearchTimeout(delay time.Duration) {
+	if t.stopSearchTimeout != nil {
+		return
+	}
+	now := time.Now().Add(delay)
+	t.stopSearchTimeout = &now
+}
+
+// isStopSearchDelayExpired returns true if the timeout to stop current topci
```
Comment (dshulyak) [inline geth/peers/topicpool.go]: typo

**Comment by adambabik** [comment]
Fixed:
- https://github.com/status-im/status-go/pull/996/commits/41027c4a1bc7c36ba67c04df91300661141937b9
- https://github.com/status-im/status-go/pull/996/commits/f9228cfada9e835c2fa9f3c42f619baf0e906402

File: geth/peers/topicpool_test.go [hunk 8]
```diff
@@ -268,3 +268,39 @@ func (s *TopicPoolSuite) TestPeerDontAddTwice() {
 	// peer2 already added to p2p server no reason to add it again
 	s.Nil(s.topicPool.AddPeerFromTable(s.peer))
 }
+
+func (s *TopicPoolSuite) TestMaxCachedPeers() {
```
Comment (dshulyak) [inline geth/peers/topicpool_test.go:279]: why do you have to drop every peer?
i think that correct test wouldn't do that. just found a peer -> confirm added
and then found 2 more. after that 3 peers must be cached

it seems that in the case that i described those 2 wont be cached because of https://github.com/status-im/status-go/blob/develop/geth/peers/topicpool.go#L393

**Review by dshulyak — CHANGES_REQUESTED** [review]

File: geth/peers/topicpool_test.go (hunk 8 above)
Comment (adriacidre) [reply geth/peers/topicpool_test.go:279]: @dshulyak but, if i understood well, entry point for adding new peers is on [p2p.PeerEventTypeAdd events](https://github.com/status-im/status-go/pull/989/files#diff-fb40a852d6fa4cad34711d85a9689097R241) which will be calling [t.ConfirmAdded](https://github.com/status-im/status-go/pull/989/files#diff-fb40a852d6fa4cad34711d85a9689097R258) for every new peer on every topic.

While `processFoundNode` will add the peer to `pendingPeers` map `ConfirmAdded` will get this map entry and add it to the cache with `t.cache.AddPeer`.

The original test was simulating this behaviour by running something like this:
```go
func (s *TopicPoolSuite) TestMaxCachedPeers() {
	s.topicPool.limits = params.NewLimits(1, 1)
	s.topicPool.maxCachedPeers = 3
	peer1 := discv5.NewNode(discv5.NodeID{1}, s.peer.Self().IP, 32311, 32311)
	peer2 := discv5.NewNode(discv5.NodeID{2}, s.peer.Self().IP, 32311, 32311)
	peer3 := discv5.NewNode(discv5.NodeID{3}, s.peer.Self().IP, 32311, 32311)
	s.topicPool.processFoundNode(s.peer, peer1)
	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer1.ID))
	s.topicPool.processFoundNode(s.peer, peer2)
	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer2.ID))
	s.topicPool.processFoundNode(s.peer, peer3)
	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer3.ID))

	cached := s.topicPool.cache.GetPeersRange(s.topicPool.topic, 5)
	s.Equal(3, len(cached))

	s.True(s.topicPool.maxCachedReached())
}
```

Is there anything i'm missing?

Comment (dshulyak) [reply geth/peers/topicpool_test.go:279]: so at first node gets discovered and processFoundNode is called

if max limit is not reached this peer will be added to a server
if handshake in the server is successful -> confirm added is called and peer is persisted
if we added more than max then surplus is dropped but we don't remove those peers from cache

if max limit is reached peer is added only to a heap queue which is not persisted

test that you pasted in the comment looks correct

Comment (dshulyak) [reply geth/peers/topicpool_test.go:279]: sorry, it is not correct
because this line https://github.com/status-im/status-go/blob/develop/geth/peers/topicpool.go#L394 won't be executed and thus ConfirmAdded won't be called

maybe it should be intrsumented with boolean to understand if peer was added to server or not

Comment (adriacidre) [reply geth/peers/topicpool_test.go:279]: @dshulyak I probably misunderstood the pendingPeers vs. connectedPeers on first instance. 

So, peers found between maxPeers and maxCachedPeers should be part of `connectedPeers` instead of `pendingPeers` as they're confirmed server peers and not peers to be confirmed. My initial thought was to think on pendingPeers as cachedPeers which is not correct.

I think [this commit](https://github.com/status-im/status-go/pull/989/commits/580f07c0b1b2ef774ccc3d012a813cf170901312) will addresses the problem.

**Comment by adriacidre** [comment]
Flaky tests report: 

```
INFO [05-30|10:14:12] Notification received                    package=status-go/signal                    event="{\"type\":\"node.stopped\",\"event\":null}"
--- FAIL: TestWhisperMailboxTestSuite (41.50s)
    --- FAIL: TestWhisperMailboxTestSuite/TestRequestMessageFromMailboxAsync (11.31s)
	Error Trace:	whisper_mailbox_test.go:121
	Error:      	Not equal: 
	            	expected: 1
	            	actual: 0
INFO [05-30|10:14:12] Status backend initialized 
```


```
--- FAIL: TestTransactionsTestSuite (134.77s)
    --- FAIL: TestTransactionsTestSuite/TestSendEther (5.80s)
	Error Trace:	transactions_test.go:378
	Error:      	Received unexpected error:
	            	no suitable peers available
	Messages:   	cannot send transaction
	Error Trace:	transactions_test.go:293
			signals.go:77
			asm_amd64.s:573
			cgocall.go:316
			cgocall.go:194
			asm_amd64.s:826
			asm_amd64.s:673
			cgocall.go:131
			signals.go:41
			events_sign.go:20
			notifications.go:32
			asm_amd64.s:2361
	Error:      	Received unexpected error:
	            	no suitable peers available
	Messages:   	cannot complete queued transaction[6a587dbb-bda4-409c-8db6-d150b75c65c1]
	Error Trace:	transactions_test.go:387
	Error:      	Should be false
	Messages:   	transaction was never queued or completed
FAIL
```


```
WARN [05-30|10:45:46] message loop                             peer=eb8258b6093fbd4b err=EOF
--- FAIL: TestTransactionsTestSuite (133.05s)
    --- FAIL: TestTransactionsTestSuite/TestEmptyToFieldPreserved (5.83s)
	Error Trace:	transactions_test.go:175
			signals.go:77
			asm_amd64.s:573
			cgocall.go:316
			cgocall.go:194
			asm_amd64.s:826
			asm_amd64.s:673
			cgocall.go:131
			signals.go:41
			events_sign.go:20
			notifications.go:32
			asm_amd64.s:2361
	Error:      	Received unexpected error:
	            	no suitable peers available
	Error Trace:	transactions_test.go:188
	Error:      	"{"jsonrpc":"2.0","id":1,"error":{"code":-32000,"message":"no suitable peers available"}}" should not contain "error"
FAIL
```

cc: @adambabik

**Comment by adriacidre** [comment]
@adambabik @dshulyak is there anything missing on this PR?

File: geth/peers/topicpool_test.go (hunk 8 above)
Comment (dshulyak) [reply geth/peers/topicpool_test.go:279]: sorry just noticed your comment

1. i am not sure why peers that were dropped now present in connectedPeers table. connectedPeers are only for peers that connected to mail server. caching them should not affect that table

2. this test shouldn't drop peers at all. dropping a peer must affect count of connected peers otherwise logic would be even less clear. when dismissed peer dropped we should remove it from connected but keep in cache


i suggest to adjust the test so that it will find 3 peers, call confirm added for all 3, 2 must be dismissed after that, but no reason to call drop. cache must have all 3 peers. connected 3 but 2 dismissed. only after that drop 2 dismissed peers - connected will be 1, and cached all 3.

File: geth/peers/topicpool.go [hunk 9]
```diff
@@ -52,7 +56,8 @@ type TopicPool struct {
 	poolWG sync.WaitGroup
 	quit   chan struct{}
 
-	running int32
+	running        int32
+	maxCachedPeers int
```
Comment (dshulyak) [inline geth/peers/topicpool.go]: i think `running` here is either 0 or 1. adding maxCachedPeers next to it might be confusing

**Review by dshulyak — APPROVED** [review]

File: geth/peers/peerpool.go [hunk 10]
```diff
@@ -37,6 +37,8 @@ const (
 	DefaultDiscV5Timeout = 3 * time.Minute
 	// DefaultTopicFastModeTimeout is a timeout after which sync mode is switched to slow mode.
 	DefaultTopicFastModeTimeout = 30 * time.Second
+	// DefaultTopicStopSearchDelay is the default delay when stopping a topic search
```
Comment (themue) [inline geth/peers/peerpool.go]: Interpunction. _nitpicking_ 😁

File: geth/peers/peerpool.go [hunk 11]
```diff
@@ -47,15 +49,19 @@ type Options struct {
 	DiscServerTimeout time.Duration
 	// AllowStop allows stopping Discovery when reaching max peers or after timeout.
 	AllowStop bool
+	// TopicStopSearchDelay time stopSearch will be waiting for max cached peers to be
+	// filled before really stopping the search.
+	TopicStopSearchDelay time.Duration
 }
 
 // NewDefaultOptions return a struct with default Options.
```
Comment (themue) [inline geth/peers/peerpool.go]: s/return/returns/

File: geth/peers/peerpool.go [hunk 12]
```diff
@@ -222,37 +229,65 @@ func (p *PeerPool) handleServerPeers(server *p2p.Server, events <-chan *p2p.Peer
...
-		} else {
+		}
+	}
+	if p.allTopicsStopped() {
+		log.Debug("closing discv5 connection because all topics reached max limit", "server", server.Self())
+		p.stopDiscovery(server)
+	}
+}
+
+// allTopicsStopped return true if all topics are stopped
```
Comment (themue) [inline geth/peers/peerpool.go]: s/return/returns/ and interpunction. I'm pedantic on comments, I know. but comments are code too.

File: geth/peers/peerpool.go [hunk 13]
```diff
@@ -222,37 +229,65 @@ func (p *PeerPool) handleServerPeers(server *p2p.Server, events <-chan *p2p.Peer
...
+func (p *PeerPool) allTopicsStopped() (all bool) {
+	if !p.opts.AllowStop {
```
Comment (themue) [inline geth/peers/peerpool.go]: Hmm, this usage shows that a naming like `AllowsStop` would match better. But not important.

File: geth/peers/peerpool.go [hunk 14]
```diff
@@ -222,37 +229,65 @@ func (p *PeerPool) handleServerPeers(server *p2p.Server, events <-chan *p2p.Peer
...
+		return false
+	}
+	all = true
+	for _, t := range p.topics {
+		if !t.stopped() {
```
Comment (themue) [inline geth/peers/peerpool.go]: Like here an `isStopped()`. Done Smalltalk for too long. 😃

File: geth/peers/topicpool.go [hunk 15]
```diff
@@ -141,6 +148,40 @@ func (t *TopicPool) BelowMin() bool {
 	return len(t.connectedPeers) < t.limits.Min
 }
 
+// maxCachedReached returns true if max number of cached peers is reached.
+func (t *TopicPool) maxCachedReached() bool {
```
Comment (themue) [inline geth/peers/topicpool.go]: `maxCachedPeersReached()`?

**Review by themue — APPROVED** [review]
Mostly LGTM beside already open comments, only few nitpickings.

File: geth/peers/peerpool.go (hunk 12 above)
Comment (adriacidre) [reply geth/peers/peerpool.go]: no worries, at least on my PRs you can be as much pedantic as you want :)
